### PR TITLE
JAVA: Expose two new methods for GDAL's JNI Bridge

### PR DESCRIPTION
**Pull request by mglazer**
## What does this PR do?

Exposes two new JNI methods. Specifically:

* VSIFileFromMemBuffer - Augments to support taking a ByteBuffer as an
  argument
* VSIGetMemFileFromBuffer - Exposes this method and returns a ByteBuffer
  pointer to the underlying memory.

## What are related issues/pull requests?

n/a, just exposing what could have been exposed from the start. This is my first gdal contribution as well as some of the first swig code I've written so apologies if I did anything incorrectly.

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: gdal/swig/include/cpl.i [hunk 1]
```diff
@@ -520,6 +521,27 @@ void wrapper_VSIFileFromMemBuffer( const char* utf8_path, int nBytes, const GByt
 #endif
 #endif
 
+#if defined(SWIGJAVA)
+%inline %{
+void wrapper_VSIFileFromMemBuffer( const char* utf8_path, void* nioBuffer, long nioBufferSize) {
```
Comment (rouault) [inline gdal/swig/include/cpl.i:526]: This method is already available in the Java case a bit above. Are we sure that the 2 can co-exist ? Adding a test program in swig/java/apps/ checking  that both work would be needed

File: gdal/swig/include/cpl.i [hunk 2]
```diff
@@ -520,6 +521,27 @@ void wrapper_VSIFileFromMemBuffer( const char* utf8_path, int nBytes, const GByt
...
+    VSIFCloseL(VSIFileFromMemBuffer(utf8_path, (GByte*) nioBuffer, nioBufferSize, FALSE));
+}
+%}
+#endif
+
+#if defined(SWIGJAVA)
+%apply (jobject outBuffer) {jobject};
+%inline %{
+jobject wrapper_VSIGetMemFileBuffer(JNIEnv *jenv, const char* utf8_path) {
```
Comment (rouault) [inline gdal/swig/include/cpl.i:535]: hum I don't think that playing directly with JNIEnv* here is the most idiomatic way of proceeding. Ideally we should minimize the amount of language specificities here. A cleaner way would probably to have something like
void wrapper_VSIGetMemFileBuffer(const char* utf8_path, void** outBuffer, size_t* outBufferSize) and a typemap for (void** outBuffer, size_t* outBufferSize).

File: gdal/swig/include/cpl.i [hunk 3]
```diff
@@ -520,6 +521,27 @@ void wrapper_VSIFileFromMemBuffer( const char* utf8_path, int nBytes, const GByt
...
+    vsi_l_offset outBufferLen = 0;
+    GByte* result = VSIGetMemFileBuffer(utf8_path, &outBufferLen, FALSE);
+    return jenv->NewDirectByteBuffer((void*) result, outBufferLen);
```
Comment (rouault) [inline gdal/swig/include/cpl.i:538]: NewDirectByteBuffer() can only work if it is guaranteed that result won't change while the ByteBuffer is in use. There is no such guarantee the /vsimem/ file could be destroyed in-between, or its size grown, causing a realloc(), which would invalidate the original buffer. A copy should be done for safety.
The test program to add mentioned above should also test that.

File: gdal/swig/include/java/typemaps_java.i [hunk 4]
```diff
@@ -62,6 +62,10 @@
 
 %typemap(javain) SWIGTYPE *DISOWN "$javaclassname.getCPtrAndDisown($javainput)"
 
+%typemap(in, numinputs=0) JNIEnv *jenv %{
```
Comment (rouault) [inline gdal/swig/include/java/typemaps_java.i:65]: that typemap and the one later should probably be removed for other ones, as mentioned above

**Review by rouault — CHANGES_REQUESTED** [review]

**Comment by mglazer** [comment]
Still working through some changes, so not quite there yet. Seems that returning a direct byte buffer really isn't the best thing for a number of reasons (the biggest being that I spend hours trying to get the typemaps to work but couldn't). So ended up just copying the array out, which hopefully is sufficient.

**Pull request closed without merging**